substrate/frame/migrations/src/lib.rs

```diff
@@ -698,6 +717,14 @@ impl<T: Config> Pallet<T> {
 		}
 
 		let max_steps = T::Migrations::nth_max_steps(cursor.index);
+
+		// If this is the first time running this migration, exec the pre-upgrade hook.
```

Comment:
The line code above `Historic::<T>::contains_key` should already prevent it from executing a migration twice, so this is implicit?

Reply:
If the migration runs multiple steps, I think this is still needed to ensure pre_upgrade isn't run again, because the migration is only added to Historic after it finishes.